### DISCUSSION TITLE
Why I cann't `cast call` HelperConfig contract after using DeployDSC

### DISCUSSION THREAD

**Discussion by EthanWalker10** [Cyfrin/foundry-full-course-cu#2545] [Q&A] [answered]
Hello ,guys  

In 'Section 3: Develop a DeFi protocol' , I have a trouble in calling a public variable in deployed contract `HelperConfig` which was deployed `DeployDSC`.   As we know, we deploy 3 contract (DecentralizedStableCoin, DSCEngine, HelperConfig) using DeployDSC. However, strange things come to me: 
```shell
ethan@ethan-virtual-machine:~/blockchain/advanced-foundry/DeFi-stablecoin/DeFi-stablecoin$ cast call 0x0165878A594ca255338adfa4d48449f69242Eb8F "test()(uint256)"
19
ethan@ethan-virtual-machine:~/blockchain/advanced-foundry/DeFi-stablecoin/DeFi-stablecoin$ cast call 0xC7f2Cf4845C6db0e1a1e91ED41Bcd0FcC1b0E141 "test()(uint256)"
Error: 
contract 0xc7f2cf4845c6db0e1a1e91ed41bcd0fcc1b0e141 does not have any code
```
As you can see above, I can `cast call` the same public variable in dsce contract but in HelperConfig contract.   

DeployDSC.s.sol:
```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.19;

import { Script } from "lib/forge-std/src/Script.sol";
import { HelperConfig } from "./HelperConfig.s.sol";
import { DecentralizedStableCoin } from "../src/DecentralizedStableCoin.sol";
import { DSCEngine } from "../src/DSCEngine.sol";

contract DeployDSC is Script {
    address[] public tokenAddresses;
    address[] public priceFeedAddresses;

    function run() external returns (DecentralizedStableCoin, DSCEngine, HelperConfig) {
        HelperConfig helperConfig = new HelperConfig(); // This comes with our mocks!

        (address wethUsdPriceFeed, address wbtcUsdPriceFeed, address weth, address wbtc, uint256 deployerKey) =
            helperConfig.activeNetworkConfig();
        tokenAddresses = [weth, wbtc];
        priceFeedAddresses = [wethUsdPriceFeed, wbtcUsdPriceFeed];

        vm.startBroadcast(deployerKey);
        DecentralizedStableCoin dsc = new DecentralizedStableCoin();
        DSCEngine dscEngine = new DSCEngine(tokenAddresses, priceFeedAddresses, address(dsc));
        dsc.transferOwnership(address(dscEngine));  // set the owner who own the DSC token
        vm.stopBroadcast();
        return (dsc, dscEngine, helperConfig);
    }
}
```
HelperConfig.s.sol:
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import { MockV3Aggregator } from "../test/mocks/MockV3Aggregator.sol";
import { Script } from "lib/forge-std/src/Script.sol";
import { ERC20Mock } from "@openzeppelin/contracts/mocks/ERC20Mock.sol";

contract HelperConfig is Script {
    NetworkConfig public activeNetworkConfig;

    uint256 public test = 19;  // test if I can call this variable
    uint8 public constant DECIMALS = 8;
    int256 public constant ETH_USD_PRICE = 2000e8;
    int256 public constant BTC_USD_PRICE = 1000e8;

    struct NetworkConfig {
        address wethUsdPriceFeed;
        address wbtcUsdPriceFeed;
        address weth;
        address wbtc;
        uint256 deployerKey;
    }

    uint256 public DEFAULT_ANVIL_PRIVATE_KEY = 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80;

    constructor() {
        if (block.chainid == 11_155_111) {
            activeNetworkConfig = getSepoliaEthConfig();
        } else {
            activeNetworkConfig = getOrCreateAnvilEthConfig();
        }
    }

    function getSepoliaEthConfig() public view returns (NetworkConfig memory sepoliaNetworkConfig) {
        sepoliaNetworkConfig = NetworkConfig({
            wethUsdPriceFeed: 0x694AA1769357215DE4FAC081bf1f309aDC325306, // ETH / USD
            wbtcUsdPriceFeed: 0x1b44F3514812d835EB1BDB0acB33d3fA3351Ee43,
            weth: 0xdd13E55209Fd76AfE204dBda4007C227904f0a81,
            wbtc: 0x8f3Cf7ad23Cd3CaDbD9735AFf958023239c6A063,
            deployerKey: vm.envUint("PRIVATE_KEY")
        });
    }

    function getOrCreateAnvilEthConfig() public returns (NetworkConfig memory anvilNetworkConfig) {
        // Check to see if we set an active network config
        if (activeNetworkConfig.wethUsdPriceFeed != address(0)) {
            return activeNetworkConfig;
        }

        vm.startBroadcast();
        MockV3Aggregator ethUsdPriceFeed = new MockV3Aggregator(DECIMALS, ETH_USD_PRICE);  // address of PriceFeed
        ERC20Mock wethMock = new ERC20Mock("WETH", "WETH", msg.sender, 1000e8);  // address of token constract

        MockV3Aggregator btcUsdPriceFeed = new MockV3Aggregator(DECIMALS, BTC_USD_PRICE);
        ERC20Mock wbtcMock = new ERC20Mock("WBTC", "WBTC", msg.sender, 1000e8);
        vm.stopBroadcast();

        anvilNetworkConfig = NetworkConfig({
            wethUsdPriceFeed: address(ethUsdPriceFeed), // ETH / USD
            weth: address(wethMock),
            wbtcUsdPriceFeed: address(btcUsdPriceFeed),
            wbtc: address(wbtcMock),
            deployerKey: DEFAULT_ANVIL_PRIVATE_KEY
        });
    }
}
```

the `test` variable defined in DSCEngine.sol is same as in HelperConfig.s.sol. I Originnally wanna call the NetworkConfig.wethUsdPriceFeed so that I can check what price in PriceFeed, but I got an error. After that I tested the public variable `test` just like things above, I fell in the error, too.

I will appreciate it if someone could help me.

**Comment by EngrPips** [comment]
Hello @EthanWalker10, Actually, I have never tried to call a `Script` contract from the command line using `cast`. I have only used `cast` with a proper contract, such as your `dsce` contract. I don't know how a script contract such as `HelperConfig` is supposed to react when you call them using cast. I have only called a script when running a process in another script. I will advise that you read the official Foundry docs https://book.getfoundry.sh/ to find out what they say about using cast on a script contract. I am eager to hear what you find out.

**Reply by EthanWalker10** [reply]
Seems that there is nothing about this issue in https://book.getfoundry.sh/reference/cast/cast-call  
But I find that struct can be returned by `cast call` [here](https://ethereum.stackexchange.com/questions/150079/how-to-use-foundry-cast-call-when-the-method-returns-a-struct), good for me.  I wanna share it with you.

**Reply by EngrPips** [reply]
Interesting. Thanks for sharing with me. Why not ask on `stack-overflow` or `ethereum-stack-exchange`? This is interesting, and it will be nice to know why using cast on a script contract doesn't work as expected.

**Reply by EthanWalker10** [reply]
I will do it latter, now I have a big big big trouble waiting for me to solve. This one is just a little single issue I met during my process on solving the big trouble.  I wanna publish it here latter.

**Reply by EngrPips** [reply]
I am looking forward to it.

**Reply by EthanWalker10** [reply]
Friend, I already have some ideas. Once I organize the entire question better and post it, I will @ you.

**Reply by EngrPips** [reply]
Now I can't wait to hear from you, hahaha. What are friend you are, buddy!

**Reply by EthanWalker10** [reply]
OMG,  the whole situation is so ridiculous and you will gonna laugh at it. 
Simply, I fork the [repo](https://github.com/Cyfrin/foundry-defi-stablecoin-cu); and then I find testing StopOnRevertInvariants.t.sol failed.   And... you could find out why, and I will tell you the whole situation and why tomorrow, cause now I have to have some sports now hahahaha

**Reply by EngrPips** [reply]
> OMG, the whole situation is so ridiculous and you will gonna laugh at it. Simply, I fork the [repo](https://github.com/Cyfrin/foundry-defi-stablecoin-cu); and then I find testing StopOnRevertInvariants.t.sol failed. And... you could find out why, and I will tell you the whole situation and why tomorrow, cause now I have to have some sports now hahahaha

I don't think I will laugh at you, I have experienced some really ridiculous moments myself.

**Reply by EthanWalker10** [reply]
GM! It's just a joke haha. I found there are 2 main errors in my testing StopOnRevertInvariants.t.sol and StopOnRevertHandler.t.sol
: 
1. `assert(wethValue + wbtcValue >= totalSupply);` is broken;
2. minted dsc often more than deposited collateral.
Actually they are a same question, one happened in StopOnRevertInvariants.t.sol and another in StopOnRevertHandler.t.sol. 
Once I thought maybe the randomness of fuzz made this. However, I had reviewed a lot of sources and haven't found any mention of it. In fact, I reviewed the mock code and believed that the price of PriceFeed can be changed only if calling the update function. It's a pity I didn't move further.  
  
Then I did many works such as writing tests for getUsdValue(), adding variables and logging them, trying to inspect the price in PriceFeed(which got failed you know), and so on...  

When I had done all these works, I was sure that only one way made this error: There must be a place where this update function was called.  And then I got it in StopOnRevertHandler.t.sol....  

Anyway, it's not wasting time for me. At first, I was still relatively unfamiliar with the mechanism of fuzz testing so that I added counters in functions to grasp it better(but I didn't notice the updateCollateral function which was a sad story). But now, I have much deeper in it and I feel good for it. Sharing it with you, my friend.

**Reply by EngrPips** [reply]
Interesting. This is basically the life of every programmer, and yeah, with time, it will get a lot easier. I am honored you shared this with me, Friend.

**Comment by EthanWalker10** [comment, marked as answer]
Now I know that: we can't `cast call` a script

**Reply by EngrPips** [reply]
Interesting.
